velox/exec/fuzzer/WindowFuzzer.cpp

```diff
@@ -68,20 +68,66 @@ void WindowFuzzer::addWindowFunctionSignatures(
   }
 }
 
-std::tuple<std::string, bool> WindowFuzzer::generateFrameClause() {
-  auto frameType = [](int value) -> const std::string {
+void WindowFuzzer::addKRangeFrameColumnToInput(
+    std::vector<RowVectorPtr>& input,
+    core::WindowNode::BoundType frameBoundType,
+    const std::string& columnName) {
+  // For frames with k RANGE PRECEDING/FOLLOWING, Velox requires the application
+  // to add columns with the range frame boundary value computed according
+  // to the frame type.
+  // If the frame is k PRECEDING :
+  // frame_boundary_value = current_order_by - k (for ascending ORDER BY)
+  // frame_boundary_value = current_order_by + k (for descending ORDER BY)
+  // If the frame is k FOLLOWING :
+  // frame_boundary_value = current_order_by + k (for ascending ORDER BY)
+  // frame_boundary_value = current_order_by - k (for descending ORDER BY)
+  auto computeOffset = [&](const core::WindowNode::BoundType& frameBoundType,
+                           const int64_t& offsetValue,
+                           const bool& ascending) -> int64_t {
+    if (frameBoundType == core::WindowNode::BoundType::kPreceding) {
+      return ascending ? -offsetValue : offsetValue;
+    } else {
+      return ascending ? offsetValue : -offsetValue;
+    }
+  };
```

Comment:
@kagamiori I think this lambda function should stay here so it can take care of k-range frames in the future where the order by column is descending and not ascending. Right now I have it always true from the call below because I'm using the row_number column as the order by column which is ascending order.

```
// Setting ascending boolean to true since we're using ascending row_number
  // column as the order by column
  int64_t offset = computeOffset(frameBoundType, offsetValue, true);
```

Reply:
Hi @minhancao, I see. Ideally the offset should be relative to the single sorting key column. Are you going to extend it in a subsequent PR?

nit: change the name `computeOffset` to `determineOffset` since this lambda doesn't really do any computation.

Reply:
Yes I will extend it in a subsequent PR for when the k-range frame is using any other order by column that's not row_number column